luminoth/models/fasterrcnn/rcnn_target.py

```diff
@@ -76,97 +68,161 @@ def _build(self, proposals, gt_boxes):
                 other proposal we return zeros.
                 The shape of the Tensor is (num_proposals, 4).
         """
-        (proposals_label, bbox_targets) = tf.py_func(
-            self.proposal_target_layer,
-            [proposals, gt_boxes],
-            [tf.float32, tf.float32],
-            stateful=False,
-            name='proposal_target_layer'
-         )
 
-        return proposals_label, bbox_targets
+        # TODO: revise casts to int64 in tf.sparse_to_dense and tf.scatter_nd
+        # calls.
 
-    def proposal_target_layer(self, proposals, gt_boxes):
-        """
-        Numpy implementation for _build.
-        """
-        if self._debug:
-            np.random.seed(0)  # TODO: For reproducibility.
-
-        # Remove batch id from proposals
+        # Remove batch id from proposals.
         proposals = proposals[:, 1:]
 
-        overlaps = bbox_overlap(proposals, gt_boxes)
-
-        # overlaps returns (num_proposals, num_gt_boxes) with the IoU of
+        overlaps = bbox_overlap_tf(proposals, gt_boxes[:, :4])
+        # overlaps now contains (num_proposals, num_gt_boxes) with the IoU of
         # proposal P and ground truth box G in overlaps[P, G]
 
-        # We are going to label each proposal based on the IoU with `gt_boxes`.
-        # Start by filling the labels with -1, marking them as ignored.
-        proposals_label = np.empty((proposals.shape[0], ), dtype=np.float32)
-        proposals_label.fill(-1)
-
+        # We are going to label each proposal based on the IoU with
+        # `gt_boxes`. Start by filling the labels with -1, marking them as
+        # ignored.
+        proposals_label_shape = tf.gather(tf.shape(proposals), [0])
```

Comment:
The difference seems to be that `[tf.shape(proposals)[0]]` gives a python `list` of `Tensor`s. (While `Tensor.shape` returns a python `Object`, `tf.shape` can't, because it runs on the graph. So it returns a `Tensor`). On the other hand, `tf.gather(tf.shape(proposals), [0])` returns a single `Tensor`.

Surprisingly enough, `tf.stack([tf.shape(proposals)[0]])` doesn't work either, so there may be more to this than meets the eye.